pkg/manager/tidbngmonitoring/ng_monitoring_manager.go

```diff
@@ -0,0 +1,503 @@
+	// downward
+	builder.PodTemplateSpecBuilder().ContainerBuilder(nmContainerName).AddEnvs(spec.Env()...)
+	builder.PodTemplateSpecBuilder().AddLabels(spec.Labels())
+	builder.PodTemplateSpecBuilder().AddAnnotations(spec.Annotations())
+	// additional storage volumes
+	storageVolMounts, additionalPVCs := util.BuildStorageVolumeAndVolumeMount(tngm.Spec.NGMonitoring.StorageVolumes, tngm.Spec.NGMonitoring.StorageClassName, v1alpha1.NGMonitoringMemberType)
+	builder.PodTemplateSpecBuilder().ContainerBuilder(nmContainerName).AddVolumeMounts(storageVolMounts...)
+	builder.AddVolumeClaims(additionalPVCs...)
+	// additional volumes and mounts
+	builder.PodTemplateSpecBuilder().ContainerBuilder(nmContainerName).AddVolumeMounts(spec.AdditionalVolumeMounts()...)
+	builder.PodTemplateSpecBuilder().AddVolumes(spec.AdditionalVolumes()...)
+	// additional containers
+	builder.PodTemplateSpecBuilder().AddContainers(spec.AdditionalContainers()...)
+	// TODO: TLS
+
+	return builder.Get(), nil
+}
+
+// GenerateNGMonitoringConfigMap generate ConfigMap from tidb ng monitoring
+func GenerateNGMonitoringConfigMap(tngm *v1alpha1.TidbNGMonitoring) (*corev1.ConfigMap, error) {
+	config := tngm.Spec.NGMonitoring.Config
+	meta, _ := GenerateNGMonitoringMeta(tngm, NGMonitoringName)
+
+	// TODO: TLS
+
+	confText, err := config.MarshalTOML()
+	if err != nil {
+		return nil, err
+	}
+
+	name := NGMonitoringName(tngm.Name)
+	confTextStr := string(confText)
+
+	data := map[string]string{
+		ngmConfigMapConfigKey: confTextStr,
+	}
+
+	meta.Name = name
+
+	return &corev1.ConfigMap{
+		ObjectMeta: meta,
+		Data:       data,
+	}, nil
+}
+
+// GenerateNGMonitoringMeta build ObjectMeta and Label for ng monitoring
+func GenerateNGMonitoringMeta(tngm *v1alpha1.TidbNGMonitoring, nameFunc func(string) string) (metav1.ObjectMeta, label.Label) {
+	instanceName := tngm.GetInstanceName()
+	label := label.NewTiDBNGMonitoring().Instance(instanceName).NGMonitoring()
+
+	objMeta := metav1.ObjectMeta{
+		Name:            nameFunc(tngm.Name),
+		Namespace:       tngm.GetNamespace(),
+		Labels:          label,
+		OwnerReferences: []metav1.OwnerReference{controller.GetTiDBNGMonitoringOwnerRef(tngm)},
+	}
+	return objMeta, label
+}
+
+// GenerateNGMonitoringHeadlessService build headless service for ng monitoring
+func GenerateNGMonitoringHeadlessService(tngm *v1alpha1.TidbNGMonitoring) *corev1.Service {
+	meta, labels := GenerateNGMonitoringMeta(tngm, NGMonitoringHeadlessServiceName)
+
+	return &corev1.Service{
+		ObjectMeta: meta,
+		Spec: corev1.ServiceSpec{
+			ClusterIP: "None",
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "ng-monitoring",
+					Port:       ngmServicePort,
+					TargetPort: intstr.FromInt(ngmServicePort),
+					Protocol:   corev1.ProtocolTCP,
+				},
+			},
+			Selector:                 labels,
+			PublishNotReadyAddresses: true,
+		},
+	}
+}
+
+func GenerateNGMonitoringStartScript(tngm *v1alpha1.TidbNGMonitoring) (string, error) {
+	if len(tngm.Spec.Clusters) < 1 {
```

Comment:
This should be validated before the sync starts.

Reply:
done